src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodSymbol.cs

```diff
@@ -220,21 +165,50 @@ private void MethodChecks(MethodDeclarationSyntax syntax, Binder withTypeParamsB
 
                     IReadOnlyDictionary<TypeParameterSymbol, bool> isValueTypeOverride = null;
                     declaredConstraints = signatureBinder.WithAdditionalFlags(BinderFlags.GenericConstraintsClause | BinderFlags.SuppressConstraintChecks).
-                                              BindTypeParameterConstraintClauses(this, _typeParameters, syntax.TypeParameterList, syntax.ConstraintClauses,
+                                              BindTypeParameterConstraintClauses(this, TypeParameters, syntax.TypeParameterList, syntax.ConstraintClauses,
                                                                                  ref isValueTypeOverride,
                                                                                  diagnostics, isForOverride: true);
                 }
 
                 // Force resolution of nullable type parameter used in the signature of an override or explicit interface implementation
                 // based on constraints specified by the declaration.
-                foreach (var param in _lazyParameters)
+                foreach (var param in parameters)
                 {
                     forceMethodTypeParameters(param.TypeWithAnnotations, this, declaredConstraints);
                 }
 
-                forceMethodTypeParameters(_lazyReturnType, this, declaredConstraints);
+                forceMethodTypeParameters(returnType, this, declaredConstraints);
             }
 
+
+            return (returnType, parameters, _lazyIsVararg, declaredConstraints);
+
+            static void forceMethodTypeParameters(TypeWithAnnotations type, SourceOrdinaryMethodSymbol method, ImmutableArray<TypeParameterConstraintClause> declaredConstraints)
+            {
+                type.VisitType(null, (type, args, unused2) =>
+                {
+                    if (type.DefaultType is TypeParameterSymbol typeParameterSymbol && typeParameterSymbol.DeclaringMethod == (object)args.method)
+                    {
+                        if (!args.declaredConstraints.IsDefault &&
+                            (args.declaredConstraints[typeParameterSymbol.Ordinal].Constraints & TypeParameterConstraintKind.ReferenceType) != 0)
+                        {
+                            type.TryForceResolveAsNullableReferenceType();
+                        }
+                        else
+                        {
+                            type.TryForceResolveAsNullableValueType();
+                        }
+                    }
+                    return false;
+                }, typePredicate: null, arg: (method, declaredConstraints), canDigThroughNullable: false, useDefaultType: true);
+            }
+        }
+
+        protected override void ExtensionMethodChecks(DiagnosticBag diagnostics)
+        {
+            var syntax = GetSyntax();
```

Comment:
Should these locals be declared inside the `if (IsExtensionMethod) {` block instead?